Kafka Backup and Restore Blueprint (#840)
kanisterio/kanister: examples/kafka/adobe-s3-connector/README.md

$ kubectl apply -f 'https://strimzi.io/install/latest?namespace=kafka' -n kafka

# Provision the Apache kafka and zookeeper.
$ kubectl create -f ./kafka-cluster.yaml -n kafka

# wait for the pods to be in ready state
$ kubectl wait kafka/my-cluster --for=condition=Ready --timeout=300s -n kafka

# setup kafdrop for monitoring the kafka cluster
$ kubectl create -f kafdrop.yaml -n kafka

# by default kafdrop run on port 9000, we can view it by
kubectl port-forward kafdrop 7000:9000 -n kafka
```

## Validate producer and consumer
Create Producer and Consumer using kafka image provided by strimzi
```bash
# create a producer and push data to it
$ kubectl -n kafka run kafka-producer -ti --image=strimzi/kafka:0.20.0-kafka-2.6.0 --rm=true --restart=Never -- bin/kafka-console-producer.sh --broker-list my-cluster-kafka-bootstrap:9092 --topic blogpost
> event1
> event2
> event3

# creating a consumer on a different terminal
$ kubectl -n kafka run kafka-consumer -ti --image=strimzi/kafka:0.20.0-kafka-2.6.0 --rm=true --restart=Never -- bin/kafka-console-consumer.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --topic my-topic --from-beginning
```

**NOTE:**

> * Here we have now kafka running with the broker running on service `my-cluster-kafka-bootstrap:9092`

Reviewer: can we please put the options in tabular format and tell users what the property means?

Author: changes are done.

# Kafka topic backup and restore
To Backup and restore Kafka topic data, we have used Adobe S3 Kafka connector which periodically polls data from Kafka and in turn uploads it to S3. Each chunk of data is represented as an S3 object. If no partitioner is specified in the configuration, the default partitioner which preserves Kafka partitioning is used.

While restoration, Topics messages are first purged and then restore operation is performed

## Prerequisites

* Kubernetes 1.9+

> * Kanister controller version 0.50.0 installed in your cluster, let's say in namespace <kanister-operator-namespace>. in our case we have used `kasten-io` namespace

Reviewer 2: ```suggestion
* Kanister controller version 0.50.0 installed in your cluster, let's say in namespace <kanister-operator-namespace>. In our case we have used `kasten-io` namespace
```

* `adobe-s3-sink.properties` file contain properties related `s3 sink Connector`
* `adobe-s3-source.properties` file contain properties related `s3 source Connector`
* `kafkaConfiguration.properties` contain properties pointing to kafka server

## Configuration

The following configuration applies to source and sink connector.

| Config Key | Notes |
| ---------- | ----- |
| name | name of the connector |
| s3.bucket | The name of the bucket to write too. |
| s3.prefix | Prefix added to all object keys stored in bucket to "namespace" them. |
| s3.path_style | Force path-style access to bucket |
| topics | Comma seperated list of topics that need to processed |
| task.max | max task that should be run inside a connector |
| format | S3 File Format |
| compressed_block_size | How much _uncompressed_ data to write to the file before we rol to a new block/chunk |

These additional configs apply to the kafka-connect:

| Config Key | Notes |
| ---------- | ----- |
| bootstrap.servers | Kafka broker address in the cluster |
| plugin.path | connector jar location |

## Setup Blueprint, configMap and location profile

> Before Setting up Blueprint, a profile is created which has s3 Details, alongwith that a configMap with the configuration details. `timeinSeconds` denote the time after which sink connector need to stop if running.

Reviewer 2: ```suggestion
Before Setting up Blueprint, a profile is created which has s3 Details, along with that a configMap with the configuration details. `timeinSeconds` denote the time after which sink connector need to stop if running.
```

> Before Setting up Blueprint, a profile is created which has s3 Details, alongwith that a configMap with the configuration details. `timeinSeconds` denote the time after which sink connector need to stop if running.

Reviewer 2: ```suggestion
Before Setting up Blueprint, a profile is created which has s3 Details, alongwith that a configMap with the configuration details. `timeinSeconds` denotes the time after which sink connector need to stop if running.
```

>{"title":"The Matrix","year":1999,"cast":["Keanu Reeves","Laurence Fishburne","Carrie-Anne Moss","Hugo Weaving","Joe Pantoliano"],"genres":["Science Fiction"]}
>{"title":"ABCD3","year":2000,"cast":["Keanu Reeves","Laurence Fishburne","Carrie-Anne Moss","Hugo Weaving","Joe Pantoliano"],"genres":["Science Fiction"]}
>{"title":"Student of the year","year":2001,"cast":["Keanu Reeves","Laurence Fishburne","Carrie-Anne Moss","Hugo Weaving","Joe Pantoliano"],"genres":["Science Fiction"]}
>{"title":"ABCD","year":2002,"cast":["Keanu Reeves","Laurence Fishburne","Carrie-Anne Moss","Hugo Weaving","Joe Pantoliano"],"genres":["Science Fiction"]}
```
* check S3 bucket for the topic

## Perform Restore
To perform restore, a prehook restore operation is performed which will purge all events from the topics in the kafka cluster whose backups were performed previously.
```bash

$ kanctl create actionset --action restore --namespace kasten-io --blueprint kafka-blueprint --profile kafka/s3-profile-fn64h --objects v1/configmaps/kafka/s3config

```
**NOTE:**

> * Here the topic need to be already present in the kafka cluster.

Reviewer 2: ```suggestion
* Here the topic needs to be already present in the kafka cluster.
```

* Before running prehook operation confirm that no other consumer is consuming data from that topic

## Verify restore
Create a consumer for topics
```bash
# creating a consumer on a different terminal
$ kubectl -n kafka-test run kafka-consumer -ti --image=strimzi/kafka:0.20.0-kafka-2.6.0 --rm=true --restart=Never -- bin/kafka-console-consumer.sh --bootstrap-server my-cluster-kafka-external-bootstrap:9094 --topic blogpost --from-beginning
```
All the messages restored can be viewed

## Delete Blueprint and Profile CR

```bash
# delete the blueprint
$ kubectl delete blueprints.cr.kanister.io <blueprint-name> -n kasten-io
# Get the profile
$ kubectl get profiles.cr.kanister.io -n kafka-test
NAME               AGE
s3-profile-fn64h   2h
# Delete the profile
$ kubectl delete profiles.cr.kanister.io s3-profile-fn64h -n kafka-test
```

### Troubleshooting

> If you run into any issues with the above commands,
> 
> you can check the logs of the controller using:

Reviewer 3: ```suggestion
The following debug commands can be used to troubleshoot issues during the backup and restore processes:

Check Kanister controller logs:
```

```bash
$ kubectl --namespace kasten-io logs -l run=kanister-svc -f
```

> you can check events of the actionset:

Reviewer 3: ```suggestion
Check events of the ActionSet:
```

```bash
$ kubectl describe actionset <actionset-name> -n kasten-io
```

> you can also check the logs of kanister job

Reviewer 3: ```suggestion
Check the logs of the Kanister job
```

```bash

> # get the pod name
> $ kubectl get pod -n kafka-test
> 
> # check the logs
> $ kubectl logs <name-of-pod-running the job> -n kafka-test

Reviewer 3: ```suggestion
# Get the Kanister job pod name
$ kubectl get pod -n kafka-test

# Check the logs
$ kubectl logs <name-of-pod-running the job> -n kafka-test
```
